vermouth/ffinput.py

```diff
@@ -51,196 +55,337 @@
+
+        if type(self.current_link) != type(None):
+           self.links.append(self.current_link)
+       
+    def finalize(self, lineno=0):
+
+        # super needs to go first because it calls finilze_section
+        # which appends the last links and blocks. If it does not 
+        # the blocks are not updated into the ff directory
+
+        super().finalize(lineno=lineno)
+
+        for block in self.blocks.values():
+            # Because of hos they are described in gromacs, proper and improper
+            # dihedral angles are all under the [ dihedrals ] section. However
+            # the way they are treated differently in the library, at least on how
+            # they generate edges. Here we move the all the impropers into their own
+            # [ impropers ] section.
+            propers = []
+            impropers = []
+            for dihedral in block.interactions.get('dihedrals', []):
+                if dihedral.parameters and dihedral.parameters[0] == '2':
+                    impropers.append(dihedral)
+                else:
+                    propers.append(dihedral)
+            block.interactions['dihedrals'] = propers
+            block.interactions['impropers'] = impropers
+            block.make_edges_from_interactions()
+        for link in self.links:
+            link.make_edges_from_interactions()
+        self.force_field.blocks.update(self.blocks)
+        self.force_field.links.extend(self.links)
+        modifications = {mod.name: mod for mod in self.modifications}
+        self.force_field.modifications.update(modifications)
+
+    def get_context(self, context_type):
+        possible_contexts = {
+            'block': self.current_block,
+            'link': self.current_link,
+            'molmeta': self.current_link,
+            'modification': self.current_modification,
+        }
+        return possible_contexts[context_type]
+
+    def has_context(self):
+        open_contexts = [
+            self.current_block, self.current_link, self.current_modification]
+        return open_contexts != ([None] * len(open_contexts))
+
+    def _new_block(self):
+        self.current_block = Block(force_field=self.force_field)
+
+    def _new_link(self):
+        self.current_link = Link(force_field=self.force_field)
+#        self.links.append(self.current_link)
+
+    def _new_modification(self):
+        self.current_modification = Link(force_field=self.force_field)
+        self.modifications.append(self.current_modification)
+
+    @SectionLineParser.section_parser('variables')
+    def _variables(self, line, lineno=0):
+        if self.has_context():
+            raise IOError('The [variables] section must be defined '
+                          'before the blocks, links, and modifications.')
+        tokens = _tokenize(line)
+        _parse_variables(tokens, self.force_field, 'variables')
+
+    @SectionLineParser.section_parser('moleculetype')
+    def _block(self, line, lineno=0):
```

Comment:
Move the contents of _new_block to here? I didn't check where _new_block gets called though.

Reply:
_new_block get's called as the section header is read. This parser actually parses the first line after the header. I would not move it, because the current setup is consistent with links and modifications, which cannot be moved to their respective sections  for the reason that link does not have to be followed by a line with information. Also it would somehow defy the current organization into a parser for headers and sections, where sections initialize the context and prep everything to be read.

Reply:
Ok, good point. Agreed.